R/parse_variable_indices.R

```diff
@@ -0,0 +1,148 @@
+#' Parse the indices from a vector of variables (extracted e.g. from a 
+#' `draws_summary` object or a `draws` object)
+#' 
+#' @name variable_indices
+#' @param x a character vector of variables
+#'
+#' @return
+#' A list with index information for each unique variable name V in `x`. Top-level list names are 
+#' the variable names. Each element contains: 
+#'    $dimensionality  the number of dimensions of V. Returns 0 for scalars with no brackets
+#'    but 1 for `y[1]` even if `y` has no other entries in `x`.
+#'    
+#'    $dimensions  a vector of the actual dimensions of V, as determined by the number of unique
```

Comment:
Is `dimensions` currently used for anything?

Or put another way: is there a situation in which it is clear that someone would ever want both `dimensions` and `implied_dimensions` at the same time? I asked because I think it could simplify things to just calculate and return the implied dimensions as the canonical "dimensions". Possibly with an option to only calculate the non-implied dimensions if those are desired (in which case the function can skip all the expensive calculations involved in figuring out the implied dimensions).

Reply:
I think the main use case for both would be if somebody wants to check whether they match.  Not sure that's a "real" use case though.  

In `rollup_summary` I thought it might be more useful to report the `dimensions` rather than the `implied_dimensions` since the former give a better (albeit imperfect) sense of big the rolled-up variable actually is.

TBH, for my own purposes I don't expect `dimensions` and `implied_dimensions` to ever differ, so I'll definitely follow your lead on which quantity is more useful.

Reply:
> I think the main use case for both would be if somebody wants to check whether they match. Not sure that's a "real" use case though.

Yeah I'd vote for waiting to see if this ever comes up instead of designing for it in advance.

> In rollup_summary I thought it might be more useful to report the dimensions rather than the implied_dimensions since the former give a better (albeit imperfect) sense of big the rolled-up variable actually is.

Good point, I agree. In that case, that function doesn't need any of the stuff that comes with implied_dimensions right? Then I would propose adding a flag to parse_variable_indices that either fills in the implied dimensions or not (with default `TRUE`). I'm not sure what to call it -- `fill_implied_dimensions` is too long. Maybe just `fill` or `fill_dim` or `fill_indices` or `fill_missing`? Then if `fill_missing` is `TRUE`, `dimensions` would be reported with missing indices filled in and otherwise not.